### DISCUSSION TITLE
Help with "Navigate To" Action

### DISCUSSION THREAD

**Discussion by msj1542** [Budibase/budibase#6343] [Help]
I have two forms, Sales and Line (as in line item).  

Going into the /sales/new/row form, I have a button within to add a new line item, which when clicked does the following:
1. Validate Form
2. Save Row (which saves the row in the Sales table)
3. Save Row (which creates a new row in the Line table, and adds the Sales ID to the relational sales column  i.e.  Sales Column -  {{ Action 2.Saved row._id }}   )
4. Navigate To:   /line/{{ Action 3.Saved row._id }}

The above seems to work just fine.  However, in the 2nd form, Line Items, I have a save button that I am essentially trying to do the same thing but in reverse:  validate form > save row > return previous form, however, I am having zero luck with the Navigate To function working correctly here.  My attempt at accomplishing this is as follows:

1. Validate Form
2. Save Row (updates the row with all additional added information besides the  Sales ID already included in the relational column, from the previous action)
3. Navigate To:   /sales/{{ Action 2.Saved row.sales._id }}  (I have also attempted:   /sales/{{ Action 2.Saved row.sales }}  /sales/{{ Repeater.Line.Sale._id }} ; /sales/{{ Repeater.Line.Sale. }}, all to no avail)

Action 3 here is where I am encountering issues - it is either sending me to /sales/:id where it actually says ":id" in the url instead of the Sales ID itself, which is what I would expect to be there; or it is bypassing this screen altogether and just sending me to the Home screen, OR, it is sending me to /sales/[{"_id":"ro_ta_f61f943f9d2047109a9e710f71d934f3_6e111bec57aa4e2b8c63d9f29019bddd","primaryDisplay":4567}], where it is displaying the array instead of only the ID, which results in an error.  

I hope this all makes sense, I typed it out very quickly.  But if someone has any clue what I am doing wrong here and could let me know, I would hugely appreciate it!!  

Thanks!!

**Comment by melohagan** [comment]
Hi @msj1542 

This is because relationships are always stored in an array (even if it's just one-one)

Thus you'll want to do: **Navigate To**: `/sales/{{ Action 2.Saved row.sales.0._id }}`
> Note the '0' after row.sales

If you needed the last sales relationship, you could try a JS binding:
```
let lastSaleIndex = $("Action 2.Saved row.sales").length - 1;
return "/sales/" + $("Action 2.Saved row.sales")[lastSaleIndex]._id
```

**Reply by msj1542** [reply]
I feel like I tried the 0 at some point too, but I will go back and try that again, hopefully that will yield better results than I am having!  Thank you so much for getting back with me so quickly!!  

Do you happen to know of any documentation and/or discussions (or even mind explaining yourself, if you have time) on things like this, like when a person would/wouldnt use a 0 in the formula?  Because I have seen it both ways when trying to reference examples, but have not been able to decipher the difference between the two.  

Also, along this same line, do you know of something similar that goes into more detail on JS bindings within Budibase, like something that is a bit more in depth than the documentation on the website, or even a "cheat-sheet" of sorts would be amazing. 

Thanks again for your help!!

**Reply by melohagan** [reply]
So the general case is that for <a href="https://docs.budibase.com/docs/formula#fields-from-relationships"> relationship</a> types the '0' is needed. Relationships always come back as an array so you need to specify an index for the relationship you want to access.

You may have seen it already, but I'll link it here for good measure: https://docs.budibase.com/docs/javascript

There's so much that you can do with JavaScript bindings that I'm not sure you could really have a single cheat sheet!
Is there anything in particular you would like to see a tutorial for?

**Reply by msj1542** [reply]
@melohagan thanks for getting back with me, and so sorry for the delayed reply - been a crazy week!  Thanks for that relationship link, also!  I feel like I have read that, however, reading it again just now I believe helped reinforce some of that!  Im not sure if what I am doing would be a good use case for Java bindings or not, or if I am just not fully grasping (or inputting correctly) the use of URL passing lol.  So here is specifically what I am trying to do, and maybe you can help point me in the right direction here (and If not, no big deal either lol)....

I am trying to create a simple commission calculator app (i know there is a pre made one in templates, however the existing one doesnt exactly what the same thing I am needing it to do) where, in a nutshell, a user inputs his or her sale and then the app generates generates a commission statement, plus a couple additional functional feature improvements, which I think are root of my issue(s) here.  For UI, I would like the basic user(salesperson) to have two links from the side nav:  "Record New Sale" and "My Sales" (or whatever would be more fitting - dashboard, commissions, etc.).  The My Sales screen should be fairly straight forward and shouldn't really be a cause for frustration once the other portion is built; it will have one or two cards for something like "this period comm total" and something else, then a table below with a list of all sales specific to that user.  The Record New Sale screen though is where it gets a little complicated, and even so I dont think it is "super complicated" to someone who has more experience with this than I do......."Ideally" I would like the form to be laid out something like this (which essentially would be in about three parts, sort of):

![Screenshot 2022-06-23 185155](https://user-images.githubusercontent.com/102364827/175433339-0079663d-312d-47ea-8d08-2f0550a3428c.jpg)

The save at the top would save all items within the form, the Add Line Item would dynamically recreate / duplicate everything within that Line Item box and add it below that box (in theory an infinite number of times, if one were so inclined), and then within the Line Item box (or sub-form, if you will) have the option to split the commission with one or many other employees for that specific Line Item (so a sub-sub form, or sorts).  I feel like to achieve this specific functionality exactly, some Java will be required, since I don't think this is built into Budibase, but I could be wrong.  

Being that I am not as advanced to be able to work through in my mind how to do this, at least within Budibase, I was trying to achieve a similar outcome through the use of Repeater groups and table button that navigates to a modal for the sub form, and "sub-sub form."  I.e.  Record new sale > would have two fields at the top and then a repeater group/table below (that would be empty on a new record create) and then add line item button which would navigate to either a page or modal that would have the line item fields on it as you see above, and another repeater group/table below that to take you to a 3rd page to save the employee percentage split information, then when you saved each linked page it takes you back to the previous page with the new respective information included in the table/repeater group below.  I did get this functionality to work mostly, but once I got one of my issues fixed (that I asked you about the other day), it broke another link, which I could never figure out the issue to kind of took a break with it.

My table structure includes 5 tables:  one for "Transactions" which has two columns for the ticket number and date fields you see in the above form, and then a relationship to a table "Line Items."  The Line Items table has 4 columns, also for the fields you see above, with the Items column forming a relationship to a table Items to populate a selectable list of values here, and then of course a 5th column that has a relationship to a "commission split" table which has two columns, one referencing the user table for another selectable list of values, and then the number field for percentage.

So, you see (at least I do, my my own mind anyways lol) how what I am trying to create is relatively straight forward, with the exception of dynamically adding a sub form (which may be easier too than I am thinking), and also with the table referencing, where A references B then B references C, then ultimately generating a table that is filtered out all table A rows that are related to the user referenced in table C (hopefully that last part makes sense lol).

Anyway though, that is essentially what I am working with here.  If you don't feel like taking the time to fully respond to all (or any) of this I totally understand lol (I praise you for even making it all the way down to this part haha), but if there is even any little pointers in here you see where I may be making this tougher than it needs to be, or something that you could point out, or anything in between, I would be very appreciative of any advice/help/input/etc. you are able to provide here.  Thanks again!!

**Reply by melohagan** [reply]
Hi @msj1542 

There's a fair bit there, so I probably wont catch everything, but let me share some links:

- Dynamically adding a sub form is not really possible at the moment - there's an issue open here: https://github.com/Budibase/budibase/issues/5658
- You may find Form Steps suitable for this scenario, but you don't have to use them: <a href="https://docs.budibase.com/docs/form-steps">form steps</a>. (It also shows how to save to more than one table)
- Another example of saving linked tables in one form: <a href="https://docs.budibase.com/docs/referencing#example-creating-and-linking-two-table-entries-in-one-form">example</a>
